### PR TITLE
refactor: isomorphic snapshot for CLI

### PR DESCRIPTION
**Pull request by bartlomieju**
* refactor main, compiler and web worker JS runtimes
* create "isomorphic" snapshot for main/worker runtimes by deferring global 
  variable init until isolate startup
* remove lib.webworker.d.ts

### CODE REVIEW

File: cli/build.rs [hunk 1]
```diff
@@ -14,6 +14,12 @@ fn main() {
   let c = PathBuf::from(env::var_os("CARGO_MANIFEST_DIR").unwrap());
   let o = PathBuf::from(env::var_os("OUT_DIR").unwrap());
 
+  let root_names = vec![c.join("js/worker_globals.ts")];
+  let bundle = o.join("WORKER_SNAPSHOT.js");
+  let state = deno_typescript::compile_bundle(&bundle, root_names).unwrap();
+  assert!(bundle.exists());
+  deno_typescript::mksnapshot_bundle(&bundle, state).unwrap();
```
Comment (ry) [inline cli/build.rs]: How much bigger does it make the exe? And how much of a slow down to the build time?

Comment (bartlomieju) [reply cli/build.rs]: The run snapshot is 1.2Mb, but I'm pretty sure we'll be able to get away without needing to create third snapshot. #3731 has to be done first though

**Comment by bartlomieju** [comment]
@ry I know this PR introduces unnecessary new snapshot, but I promise it's just temporary - I'd rather land a few smaller patches than one gigantic one. Anyway, I removed declaration for `lib.webworker.d.ts` and even with 3rd snapshot net exec size is smaller than current master.

Next steps are to :
- merge main and worker snapshot by providing a way to declare which global set to use during startup
- add proper types for different global scopes

File: cli/startup_data.rs [hunk 2]
```diff
@@ -32,6 +33,34 @@ pub fn deno_isolate_init() -> StartupData<'static> {
...
+  let source = "";
+
+  StartupData::Script(Script {
+    filename: "gen/cli/bundle/worker_main.js",
+    source,
+  })
+}
+
+#[cfg(not(feature = "no-snapshot-init"))]
+pub fn worker_isolate_init() -> StartupData<'static> {
```
Comment (ry) [inline cli/startup_data.rs]: Add a TODO that we want to remove the worker snapshot

File: cli/build.rs [hunk 3]
```diff
@@ -45,6 +45,26 @@ fn main() {
   let c = PathBuf::from(env::var_os("CARGO_MANIFEST_DIR").unwrap());
   let o = PathBuf::from(env::var_os("OUT_DIR").unwrap());
 
+  // Worker snapshot
```
Comment (ry) [inline cli/build.rs]: Add TODO that we want to remove worker snapshot

File: cli/js/compiler_bootstrap.ts [hunk 4]
```diff
@@ -35,11 +26,15 @@ const options = host.getCompilationSettings();
...
-/** Used to generate the foundational AST for all other compilations, so it can
- * be cached as part of the snapshot and available to speed up startup */
-export const oldProgram = ts.createProgram({
+/**
+ * This function spins up TS compiler and loads all available libraries
+ * into memory (including ones specified above).
+ *
+ * Used to generate the foundational AST for all other compilations, so it can
+ * be cached as part of the snapshot and available to speed up startup.
+ **/
```
Comment (ry) [inline cli/js/compiler_bootstrap.ts]: */

**Comment by bartlomieju** [comment]
I've changed my mind, I'll do it in one go...

**Comment by bartlomieju** [comment]
CC @kitsonk

**Pull request merged**